feat: run safenode services in user mode

jacderida opened the pull request:
- f5ed0523 **feat: run safenode services in user mode**

  The `Systemd` and `Launchd` service managers support running services in user mode, and the
  `service-manager-rs` crate already accommodates it, so we can support this scenario.

  Services are managed in user mode depending on whether the node manager is running elevated or not.
  In some cases, elevated privileges are still required, because not all service managers support
  user-mode services. For example, Alpine uses OpenRC, which does not have support. A service running
  in user-mode requires an active user session. This may not always be appropriate, particularly for
  advanced users who want to leave nodes running permanently in the background.

  The RPC daemon and the faucet services are still defined to run as system-wide services and they
  don't support running in user-mode, because it doesn't really make sense to require an active user
  session for these services.

- 0ad2e90f **docs: update cli and readme for user-mode services**

  The node manager documentation is updated in the README and the CLI for the introduction of
  user-mode services.

  The README also needed a little update for the `local` subcommand.

- 2d09deee **fix(launchpad): prevent mac opening with sudo**

  Add context to some error messages

- 5df12090 **chore: change terminal launch behaviour**

  On Linux, the `TERM` variable is not really appropriate for selecting which terminal is in use, so
  this has been removed. It can be set to a completely different value, like `xterm-256color`, which
  does not necessarily correspond to the running terminal. The `sudo` was also removed from the
  terminal launch mechanism.

  On the Windows front, I updated Windows Terminal to not use the `/c` argument. It simply just passes
  the command to the terminal.

  Also fixed another couple of issues:
  * Compile the node manager correctly on Windows.
  * The use of `<service>` in doc comments was being flagged as an unclosed HTML tag.

## Description

reviewpad:summary

davidrusu commented on sn_node/src/data/mod.rs:
```diff
@@ -0,0 +1,227 @@
+
+/// Events produced in the data module.
+#[derive(Debug, Clone)]
+pub enum Event {
```
Are we still using these events for some business logic?

oetyng replied:
I believe these are correct to keep, as to actually have the decoupling. Otherwise we introduce the full range of cmds into Data module, while those contain all sorts of things related to the node. It shouldn't be possible to call the node and ask it to do various things, from within data module, as that breaks inversion of control.

davidrusu commented on sn_node/src/node/flow_ctrl/dispatcher.rs:
```diff
@@ -49,6 +51,13 @@ impl Dispatcher {
+                } else {
+                    Ok(vec![])
+                }
+            }
```
Why don't we have `data.handle(..)` returning commands as well? Any reasons why we need to use DataEvent's here?

oetyng replied:
It shouldn't be possible to call the node and ask it to do various things, from within data module, as that breaks inversion of control.

davidrusu commented on sn_node/src/node/mod.rs:
```diff
@@ -527,20 +510,38 @@ mod core {
+                    .iter()
+                    .map(Peer::name)
+                    .collect();
+                let added: BTreeSet<_> = current.difference(&old.adults).copied().collect();
+                let removed: BTreeSet<_> = old.adults.difference(&current).copied().collect();
+                let adults_changed = !added.is_empty() || !removed.is_empty();
+
+                if adults_changed {
```
Is this not the same as:
```suggestion
                if current != old.adults {
```

oetyng replied:
Yes, I believe so. Remnant from when `added`, `removed` and `remaining` where used variables 👍

davidrusu reviewed (commented):
It's looking good!

My main question is why we need the `DataEvent`. Can we not have `Data::handle` return commands?

oetyng commented on sn_interface/src/network_knowledge/node_info.rs:
```diff
@@ -37,8 +37,12 @@ impl MyNodeInfo {
         Peer::new(self.name(), self.addr)
     }
 
+    pub fn id(&self) -> PublicKey {
```
Also I see that the public key is used unnecessarily for those flows, since old.

davidrusu reviewed (approved):
:ship:

davidrusu commented:
bors r+

joshuef commented on sn_node/benches/data_storage.rs:
```diff
@@ -103,22 +104,17 @@ fn bench_data_storage_writes(c: &mut Criterion) -> Result<()> {
             |b, (size, data_set)| {
-                let storage = get_new_data_store()
-                    .context("Could not create a temp data store")
+                let storage = get_new_register_store()
```
Can we maintain one interface for storing of data. Feels cleaner to me

joshuef reviewed (changes requested):
I think we need to break this up more. 

There's some renaming here that could go in it's own PR that'll be much easier to get in. So it'd be great to get those pulled out as a first step.

-----

Also I'm not sold on DataEvents... I think we may need to more solidly look at the API we want of a DataStorage module.

To my mind beyond the API itself it should be a black box (so that it is replaceable with some other impl) if we're going that way. Having the `/node` code need to handle `DataEvents` feels more complex than necessary.

---

w/r/t the data API. I also think we may well be wanting some sync calls. Or at least the option of sync calls from an API standpoint (thinking w/r/t returning ACKs only when data has been stored)

oetyng commented:
> w/r/t the data API. I also think we may well be wanting some sync calls. Or at least the option of sync calls from an API standpoint (thinking w/r/t returning ACKs only when data has been stored)

Yeah, I'll be revisiting this later like we discussed, so we can nail the sync aspect of ACKS.

> Also I'm not sold on DataEvents... I think we may need to more solidly look at the API we want of a DataStorage module.

Sure, we'll have to figure out what makes most sense to actually reach the decoupling.

> To my mind beyond the API itself it should be a black box (so that it is replaceable with some other impl) if we're going that way. Having the `/node` code need to handle `DataEvents` feels more complex than necessary.

Absolutely agree about the black box and being able to replace it with some other impl.
Also for usage of the actual `DataStorage` in other contexts than our network (including but not limited to; examples, tests etc.).

Events, sure that is a question of both pro's and cons.

In one aspect, our whole system is basically built on events, it's just a matter of syntax and structure.
Also when it comes to testability and debugging, cmds->events has lots of benefits. (These are actually the main ones I've been hoping that we would be able to get.)

In another aspect, there is a more feature rich and specific use of events that can - and cannot - be suitable for different systems.

The most suitable types of systems using the full fledged pattern is highly concurrent systems with many different parts concurrently operating on results from other parts. And our system doesn't actually look quite like that right now (and may never do).

***

The two common approaches to decoupling that I think of now are messaging and interfaces (i.e. traits in rust).
Since we do not use traits much, and already have a fundamentally message based system, I thought the message way was appropriate.

But the devils are in the details, so we'll have to look into the tradeoffs closer.

One way or another, I still think we really should decouple `DataStorage` from `MyNode`.

oetyng commented:
Will revisit this later.

jacderida committed:
docs: update cli and readme for user-mode services

The node manager documentation is updated in the README and the CLI for the introduction of
user-mode services.

The README also needed a little update for the `local` subcommand.

jacderida committed:
feat: run safenode services in user mode

The `Systemd` and `Launchd` service managers support running services in user mode, and the
`service-manager-rs` crate already accommodates it, so we can support this scenario.

Services are managed in user mode depending on whether the node manager is running elevated or not.
In some cases, elevated privileges are still required, because not all service managers support
user-mode services. For example, Alpine uses OpenRC, which does not have support. A service running
in user-mode requires an active user session. This may not always be appropriate, particularly for
advanced users who want to leave nodes running permanently in the background.

The RPC daemon and the faucet services are still defined to run as system-wide services and they
don't support running in user-mode, because it doesn't really make sense to require an active user
session for these services.

joshuef committed:
fix(launchpad): prevent mac opening with sudo

Add context to some error messages

jacderida committed:
chore: change terminal launch behaviour

On Linux, the `TERM` variable is not really appropriate for selecting which terminal is in use, so
this has been removed. It can be set to a completely different value, like `xterm-256color`, which
does not necessarily correspond to the running terminal. The `sudo` was also removed from the
terminal launch mechanism.

On the Windows front, I updated Windows Terminal to not use the `/c` argument. It simply just passes
the command to the terminal.

Also fixed another couple of issues:
* Compile the node manager correctly on Windows.
* The use of `<service>` in doc comments was being flagged as an unclosed HTML tag.

joshuef commented:
Nice one! 🎉
